common/ipmi/oem_1s_handler.c

```diff
@@ -0,0 +1,864 @@
+	memcpy(shiftdata, &msg->data[2], databyte);
+	jtag_shift_data(writebitlen, shiftdata, readbitlen, receivedata, lastidx);
+
+	memcpy(&msg->data[0], &receivedata[0], readbyte);
+	msg->data_len = readbyte;
+	msg->completion_code = CC_SUCCESS;
+	return;
+}
+
+#ifdef ENABLE_ASD
+__weak void OEM_1S_ASD_INIT(ipmi_msg *msg)
+{
+	if (msg->data_len != 1) {
+		msg->completion_code = CC_INVALID_LENGTH;
+		return;
+	}
+
+	if (msg->data[0] == 0x01) {
+		enable_PRDY_interrupt();
+	} else if (msg->data[0] == 0xff) {
+		disable_PRDY_interrupt();
+	} else {
+		disable_PRDY_interrupt();
+		msg->completion_code = CC_INVALID_DATA_FIELD;
+		return;
+	}
+
+	msg->data_len = 0;
+	msg->completion_code = CC_SUCCESS;
+	return;
+}
+#endif
+#endif
+
+__weak void OEM_1S_SENSOR_POLL_EN(ipmi_msg *msg)
+{
+	if (msg->data_len != 1) {
+		msg->completion_code = CC_INVALID_LENGTH;
+		return;
+	}
+
+	if (msg->data[0] == 1) {
+		enable_snr_poll();
+	} else if (msg->data[0] == 0) {
+		disable_snr_poll();
+	} else {
+		msg->completion_code = CC_INVALID_DATA_FIELD;
+		return;
+	}
+
+	msg->data_len = 0;
+	msg->completion_code = CC_SUCCESS;
+	return;
+}
+
+__weak void OEM_1S_ACCURACY_SENSNR(ipmi_msg *msg)
+{
+	/*********************************
+     * buf 0: target sensor number
+     * buf 1: read option
+     *          0: read from cache
+     *          1: read from sensor
+     * buf 2: sensor report status
+     ***********************************/
+	uint8_t status = -1, snr_num, option, snr_report_status;
+	int reading;
+
+	if (msg->data_len != 2) {
+		msg->completion_code = CC_INVALID_LENGTH;
+		return;
+	}
+
+	// following IPMI sensor status response
+	if (enable_sensor_poll) {
+		snr_report_status = 0xC0;
+	} else {
+		snr_report_status = 0x80;
+	}
+
+	snr_num = msg->data[0];
+	option = msg->data[1];
+
+	if (option == 0) {
+		if (enable_sensor_poll) {
+			status = get_sensor_reading(snr_num, &reading, get_from_cache);
+		} else {
+			status = SNR_POLLING_DISABLE;
+		}
+	} else if (option == 1) {
+		status = get_sensor_reading(snr_num, &reading, get_from_sensor);
+	}
+
+	switch (status) {
+	case SNR_READ_SUCCESS:
+		msg->data[0] = reading & 0xff;
+		msg->data[1] = 0x00;
+		msg->data[2] = snr_report_status;
+		msg->data_len = 3;
+		msg->completion_code = CC_SUCCESS;
+		break;
+	case SNR_READ_ACUR_SUCCESS:
+		msg->data[0] = (reading >> 8) & 0xff;
+		msg->data[1] = reading & 0xff;
+		msg->data[2] = snr_report_status;
+		msg->data_len = 3;
+		msg->completion_code = CC_SUCCESS;
+		break;
+	case SNR_NOT_ACCESSIBLE:
+	case SNR_INIT_STATUS:
+		msg->data[0] = 0x00;
+		msg->data[1] = 0x00;
+		// notice BMC about sensor temporary in not accessible status
+		msg->data[2] = (snr_report_status | 0x20);
+		msg->data_len = 3;
+		msg->completion_code = CC_SUCCESS;
+		break;
+	case SNR_POLLING_DISABLE:
+		// getting sensor cache while sensor polling disable
+		msg->completion_code = CC_SENSOR_NOT_PRESENT;
+		break;
+	case SNR_FAIL_TO_ACCESS:
+		// transection error
```

Comment:
~transection~ transaction?

Reply:
We have fixed it and done the forced update.
Please review it. Thanks.